### PR TITLE
channeldb+invoices: prepare invoice db for mpp

### PR DESCRIPTION
**Pull request by joostjager**
This PR refactors the invoice update logic to prepare it for new update flows that are required for mpp.

The main functional preparation is to make sure that the invoice `amtPaid` is recalculated with every update. For mpp we don't want to reflect accepted htlcs in the `amtPaid` until the invoice is no longer open.

Follow-up: #3415

### CODE REVIEW

File: invoices/invoiceregistry_test.go [hunk 1]
```diff
@@ -137,10 +155,10 @@ func TestSettleInvoice(t *testing.T) {
 	hodlChan := make(chan interface{}, 1)
 
 	// Try to settle invoice with an htlc that expires too soon.
-	event, err := registry.NotifyExitHopHtlc(
+	event, err := ctx.registry.NotifyExitHopHtlc(
 		hash, testInvoice.Terms.Value,
 		uint32(testCurrentHeight)+testInvoiceCltvDelta-1,
-		testCurrentHeight, getCircuitKey(10), hodlChan, nil,
+		testCurrentHeight, getCircuitKey(10), hodlChan, testPayload,
```
Comment (halseth) [inline invoices/invoiceregistry_test.go]: why this change from nil to `testPayload`?

File: record/mpp.go [hunk 2]
```diff
@@ -96,3 +97,7 @@ func (r *MPP) Record() tlv.Record {
 		MPPOnionType, r, size, MPPEncoder, MPPDecoder,
 	)
 }
+
+func (r *MPP) String() string {
```
Comment (halseth) [inline record/mpp.go]: exported methods need godoc

File: invoices/invoiceregistry.go [hunk 3]
```diff
@@ -1043,3 +955,114 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
 
 	delete(i.hodlReverseSubscriptions, subscriber)
 }
+
+// invoiceUpdater is an object that is responsible for updating the invoice
+// state and storing actions that need to be carried out after the invoice is
+// updated.
+type invoiceUpdater struct {
+	// update input fields
```
Comment (halseth) [inline invoices/invoiceregistry.go]: wip?

File: invoices/invoiceregistry.go [hunk 4]
```diff
@@ -1043,3 +955,114 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
...
+// settlement logic.
+func (i *invoiceUpdater) update(inv *channeldb.Invoice) (
+	*channeldb.InvoiceUpdateDesc, error) {
+
+	// Don't update the invoice when this is a replayed htlc.
+	htlc, ok := inv.Htlcs[i.circuitKey]
+	if ok {
+		switch htlc.State {
+		case channeldb.HtlcStateCanceled:
+			i.result = "replayed htlc to canceled invoice"
```
Comment (halseth) [inline invoices/invoiceregistry.go]: this is a bit awkward. Could we just debug log here instead?

File: invoices/invoiceregistry.go [hunk 5]
```diff
@@ -439,113 +439,25 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
...
 	// returned to the link or contract resolver.
-	invoice, err := i.cdb.UpdateInvoice(rHash, updateInvoice)
+	updater := invoiceUpdater{
+		circuitKey:           circuitKey,
+		amtPaid:              amtPaid,
+		expiry:               expiry,
+		currentHeight:        currentHeight,
+		finalCltvRejectDelta: i.finalCltvRejectDelta,
+	}
+	invoice, err := i.cdb.UpdateInvoice(rHash, updater.update)
```
Comment (halseth) [inline invoices/invoiceregistry.go]: nit: newline above

File: invoices/invoiceregistry.go [hunk 6]
```diff
@@ -439,113 +439,25 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
...
-		updateSubscribers = true
-
-		return &update, nil
-	}
-
 	// We'll attempt to settle an invoice matching this rHash on disk (if
 	// one exists). The callback will set the resolution action that is
 	// returned to the link or contract resolver.
-	invoice, err := i.cdb.UpdateInvoice(rHash, updateInvoice)
+	updater := invoiceUpdater{
```
Comment (halseth) [inline invoices/invoiceregistry.go]: personal style preference: I think encapsulating these in a struct hides a little bit what is going on with the callback. What do you think about something like

```go
invoice, err := i.cdb.UpdateInvoice(
    rHash, InvoiceUpdateCallback  {
     return update( ... params ....)
})
```

File: channeldb/invoices.go [hunk 7]
```diff
@@ -316,10 +316,12 @@ type InvoiceUpdateDesc struct {
 	// State is the new state that this invoice should progress to.
 	State ContractState
 
-	// Htlcs describes the changes that need to be made to the invoice htlcs
-	// in the database. Htlc map entries with their value set should be
-	// added. If the map value is nil, the htlc should be canceled.
-	Htlcs map[CircuitKey]*HtlcAcceptDesc
+	// CancelHtlcs describes the htlcs that need to be canceled.
```
Comment (halseth) [inline channeldb/invoices.go]: add rationale for the change to commit msg?

File: channeldb/invoices.go [hunk 8]
```diff
@@ -314,16 +314,22 @@ type HtlcAcceptDesc struct {
 // invoice.
 type InvoiceUpdateDesc struct {
 	// State is the new state that this invoice should progress to.
-	State ContractState
```
Comment (halseth) [inline channeldb/invoices.go]: `NextState`?

File: channeldb/invoices.go [hunk 9]
```diff
@@ -314,16 +314,22 @@ type HtlcAcceptDesc struct {
...
 
 	// AddHtlcs describes the newly accepted htlcs that need to be added to
 	// the invoice.
 	AddHtlcs map[CircuitKey]*HtlcAcceptDesc
+}
+
+// InvoiceStateUpdateDesc describes an invoice-level state transition.
+type InvoiceStateUpdateDesc struct {
+	// NewState is the new state that this invoice should progress to.
+	NewState ContractState
```
Comment (halseth) [inline channeldb/invoices.go]: `NextState`

File: channeldb/invoices.go [hunk 10]
```diff
@@ -314,16 +314,22 @@ type HtlcAcceptDesc struct {
...
 	// CancelHtlcs describes the htlcs that need to be canceled.
 	CancelHtlcs map[CircuitKey]struct{}
 
 	// AddHtlcs describes the newly accepted htlcs that need to be added to
 	// the invoice.
 	AddHtlcs map[CircuitKey]*HtlcAcceptDesc
+}
+
+// InvoiceStateUpdateDesc describes an invoice-level state transition.
+type InvoiceStateUpdateDesc struct {
```
Comment (halseth) [inline channeldb/invoices.go]: I don't really see why these two fields have to go into a new struct

File: channeldb/invoices.go [hunk 11]
```diff
@@ -1243,78 +1247,68 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
...
-	for key := range update.CancelHtlcs {
-		htlc, ok := invoice.Htlcs[key]
-		if !ok {
-			return nil, fmt.Errorf("unknown htlc %v", key)
-		}
-		if htlc.State != HtlcStateAccepted {
-			return nil, fmt.Errorf("can only cancel " +
-				"accepted htlcs")
+	// Update invoice state.
+	if update.State != nil {
```
Comment (halseth) [inline channeldb/invoices.go]: when do we expect this to be nil? Should at least update the doc to explain.

File: channeldb/invoices.go [hunk 12]
```diff
@@ -1243,78 +1247,68 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
...
 
 	// Process add actions from update descriptor.
 	for key, htlcUpdate := range update.AddHtlcs {
-		htlc, ok := invoice.Htlcs[key]
-
-		// Add new htlc paying to the invoice.
-		if ok {
-			return nil, fmt.Errorf("htlc %v already exists", key)
+		if _, exists := invoice.Htlcs[key]; exists {
+			return nil, errors.New("duplicate add")
```
Comment (halseth) [inline channeldb/invoices.go]: keep error message as is?

File: channeldb/invoices.go [hunk 13]
```diff
@@ -1327,6 +1321,109 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
 	return &invoice, nil
 }
 
+// updateInvoiceState validates and processes an invoice state update.
+func updateInvoiceState(invoice *Invoice, hash lntypes.Hash,
```
Comment (halseth) [inline channeldb/invoices.go]: 👍

File: channeldb/invoices.go [hunk 14]
```diff
@@ -1243,78 +1247,68 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
...
+		err := updateInvoiceState(&invoice, hash, *update.State)
+		if err != nil {
+			return nil, err
 		}
 
-		htlc.State = HtlcStateCanceled
-		htlc.ResolveTime = now
-		invoice.AmtPaid -= htlc.Amt
+		if update.State.NewState == ContractSettled {
+			err := setSettleFields(
```
Comment (halseth) [inline channeldb/invoices.go]: it is a bit confusing that both `updateInvoiceState` and `setSettleFields` is uptating the state. Could we make the first only validate the state update without altering the invoice?

File: channeldb/invoices.go [hunk 15]
```diff
@@ -1327,6 +1321,109 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
...
+		if update.Preimage.Hash() != hash {
+			return errors.New("preimage does not match")
+		}
+
+		switch invoice.State {
+		case ContractSettled:
+			return errors.New("invoice already settled")
+
+		case ContractCanceled:
+			return errors.New("invoice already canceled")
```
Comment (halseth) [inline channeldb/invoices.go]: use errors like `channeldb.ErrInvoiceAlreadyCanceled`

File: invoices/invoiceregistry.go [hunk 16]
```diff
@@ -574,32 +576,11 @@ func (i *InvoiceRegistry) CancelInvoice(payHash lntypes.Hash) error {
...
-			}
-
-			canceledHtlcs[key] = struct{}{}
-		}
-
 		// Move invoice to the canceled state.
 		return &channeldb.InvoiceUpdateDesc{
-			CancelHtlcs: canceledHtlcs,
-			State:       channeldb.ContractCanceled,
+			State: &channeldb.InvoiceStateUpdateDesc{
```
Comment (halseth) [inline invoices/invoiceregistry.go]: is the above state validation needed now?

File: channeldb/invoices.go [hunk 17]
```diff
@@ -1243,78 +1247,68 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
...
-		invoice.Terms.PaymentPreimage = update.Preimage
-
-		// Settle all accepted htlcs.
-		for _, htlc := range invoice.Htlcs {
-			if htlc.State != HtlcStateAccepted {
-				continue
+	// Align htlc states with invoice state and recalculate amount paid.
+	var amtPaid lnwire.MilliSatoshi
+	for key, htlc := range invoice.Htlcs {
+		_, cancel := update.CancelHtlcs[key]
```
Comment (halseth) [inline channeldb/invoices.go]: this reordering changes the behavior slightly, by now being able the cancel htlcs in the same update as adding them. Is that intentional?

File: invoices/invoiceregistry_test.go (hunk 1 above)
Comment (joostjager) [reply invoices/invoiceregistry_test.go]: Added commit message

File: record/mpp.go (hunk 2 above)
Comment (joostjager) [reply record/mpp.go]: added

File: invoices/invoiceregistry.go (hunk 6 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: There are quite a few parameters (another one will be added for mpp later) and also two return values atm. It could be rewritten, but I favor this approach.

File: invoices/invoiceregistry.go (hunk 3 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: no, not wip. it is a 'group comment'. I don't want to describe all these obvious fields again.

File: invoices/invoiceregistry.go (hunk 4 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: Then I need to pass in another parameter (the payment hash), to keep the log line the same. I think it is no problem to leave it as is. @carlaKC is working on exposing these failure reasons and this also prepares for that.

File: channeldb/invoices.go (hunk 8 above)
Comment (joostjager) [reply channeldb/invoices.go]: I think the type name `UpdateDesc` already indicates that this is about the next.

File: channeldb/invoices.go (hunk 7 above)
Comment (joostjager) [reply channeldb/invoices.go]: added

File: channeldb/invoices.go (hunk 10 above)
Comment (joostjager) [reply channeldb/invoices.go]: To allow the `StateUpdate` to be nil, meaning no state update.

File: channeldb/invoices.go (hunk 9 above)
Comment (joostjager) [reply channeldb/invoices.go]: New, Next, does it matter?

File: channeldb/invoices.go (hunk 11 above)
Comment (joostjager) [reply channeldb/invoices.go]: Added explanation here and at the definition of the nillable `State` field.

File: channeldb/invoices.go (hunk 14 above)
Comment (joostjager) [reply channeldb/invoices.go]: Removed duplicate state updated and renamed function name.

File: channeldb/invoices.go (hunk 12 above)
Comment (joostjager) [reply channeldb/invoices.go]: I removed the hash, because it is already logged at the call site.

File: channeldb/invoices.go (hunk 17 above)
Comment (joostjager) [reply channeldb/invoices.go]: I didn't realize that. Check added.

File: channeldb/invoices.go (hunk 15 above)
Comment (joostjager) [reply channeldb/invoices.go]: done

File: invoices/invoiceregistry.go (hunk 16 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: No not anymore, removed.

File: invoices/invoiceregistry_test.go [hunk 18]
```diff
@@ -46,19 +50,27 @@ var (
...
-func newTestContext(t *testing.T) (*InvoiceRegistry, func()) {
+type testContext struct {
+	registry *InvoiceRegistry
+
+	cleanup func()
+	t       *testing.T
+}
+
+func newTestContext(t *testing.T) *testContext {
+
```
Comment (carlaKC) [inline invoices/invoiceregistry_test.go]: unnecessary newline?

File: invoices/invoiceregistry.go (hunk 3 above)
Comment (carlaKC) [reply invoices/invoiceregistry.go]: Perhaps move the group comment to the comment on the struct then? It sticks out a bit relative to our regular comment style.

File: invoices/invoiceregistry.go (hunk 4 above)
Comment (carlaKC) [reply invoices/invoiceregistry.go]: Could we go straight to an enum for `result`? I'd have to string match or replace with an enum as is. 
I'll also need `NotifyExitHopHtlc` to return the outcome of the update, but that's less relevant for this PR.

File: invoices/invoiceregistry.go (hunk 6 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: Reconsidering this again. Hold off re-review

**Comment by joostjager** [comment]
Ok, I need to restructure this PR quite heavily. Will let know when ready again for review

File: invoices/invoiceregistry.go (hunk 3 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: Changed this part to a `invoiceUpdateCtx` and pass back the outcome as return values.

File: invoices/invoiceregistry.go (hunk 4 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: Enum created

File: invoices/invoiceregistry.go (hunk 6 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: I came around to this approach. Updated.

**Comment by joostjager** [comment]
@carlaKC @halseth this pr is ready again for re-review

File: invoices/invoiceregistry.go [hunk 19]
```diff
@@ -1043,3 +972,147 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
...
+			return nil, resultReplayToCanceled
+
+		case channeldb.HtlcStateAccepted:
+			return nil, resultReplayToAccepted
+
+		case channeldb.HtlcStateSettled:
+			return nil, resultReplayToSettled
+		}
+
+		panic("unexpected htlc state")
```
Comment (carlaKC) [inline invoices/invoiceregistry.go]: Why is this being upgraded from an error to a panic?
More for my own understanding of when we panic vs error in lnd.

File: channeldb/invoices.go [hunk 20]
```diff
@@ -1332,7 +1332,112 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
...
+	case ContractAccepted:
+		if invoice.State != ContractOpen {
+			return errors.New(
+				"can only accept open invoices")
+		}
+
+	case ContractSettled:
+		// Validate preimage.
+		if update.Preimage.Hash() != hash {
+			return errors.New("preimage does not match")
```
Comment (carlaKC) [inline channeldb/invoices.go]: Define these errors as done for `ErrInvoiceAlreadyAccepted` and update tests to check specific error while you're here?

File: invoices/invoiceregistry.go (hunk 19 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: It is a good question. I am not too sure myself either. If you for example do an index out of bounds, there will be a panic. This is to me kind of a similar situation. But if we want to be more defensive, we can return an error.

File: invoices/invoiceregistry.go [hunk 21]
```diff
@@ -1043,3 +972,147 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
...
+// update to be carried out.
+type invoiceUpdateCtx struct {
+	circuitKey           channeldb.CircuitKey
+	amtPaid              lnwire.MilliSatoshi
+	expiry               uint32
+	currentHeight        int32
+	finalCltvRejectDelta int32
+}
+
+// update is a callback for DB.UpdateInvoice that contains the invoice
```
[the remark this answers is not in the document]
Comment (joostjager) [reply invoices/invoiceregistry.go]: fixed

File: channeldb/invoices.go (hunk 20 above)
Comment (joostjager) [reply channeldb/invoices.go]: Errors defined. But where did you see these errors being checked on?

File: invoices/invoiceregistry.go (hunk 19 above)
Comment (carlaKC) [reply invoices/invoiceregistry.go]: What I mean is, why is it changed from an error to a panic in this PR? 
Because in master it just returns an [error](https://github.com/lightningnetwork/lnd/blob/cb2b8b451326d79808697af0d7f1d8d56e97dc07/invoices/invoiceregistry.go#L462)? 

Unless this branch is based on something other than master? Or I'm comparing this to the wrong code?

File: channeldb/invoices.go (hunk 12 above)
Comment (joostjager) [reply channeldb/invoices.go]: @cfromknecht made the same comment. Looking at it again, this particular thing isn't logged at the call site. Restored error msg.

File: invoices/invoiceregistry.go (hunk 19 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: There was no reason for the `error` return value anymore except for this. Then the question is, is it worth the extra code?

File: channeldb/invoices.go (hunk 20 above)
Comment (joostjager) [reply channeldb/invoices.go]: This is also a more fundamental issue. When to define an error var and when just return it in-line? In other places in this file, we still create in-line errors. Some of which also have parameters. That can all be structured too, but without that specific error being matched, it may be just boiler plate?

File: invoices/invoiceregistry.go (hunk 19 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: Brought back the `error` and leaving this discussion for another time.

File: channeldb/invoice_test.go [hunk 22]
```diff
@@ -686,7 +686,7 @@ func getUpdateInvoice(amt lnwire.MilliSatoshi) InvoiceUpdateCallback {
 		update := &InvoiceUpdateDesc{
 			Preimage: invoice.Terms.PaymentPreimage,
 			State:    ContractSettled,
-			Htlcs: map[CircuitKey]*HtlcAcceptDesc{
+			AddHtlcs: map[CircuitKey]*HtlcAcceptDesc{
```
[the remark this answers is not in the document]
Comment (joostjager) [reply channeldb/invoice_test.go]: I don't think so. Test added

File: invoices/invoiceregistry_test.go (hunk 18 above)
Comment (joostjager) [reply invoices/invoiceregistry_test.go]: hmm, removed again. wondering if there is something weird in my rebase aliases

File: invoices/invoiceregistry.go [hunk 23]
```diff
@@ -439,111 +436,46 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
...
+			*channeldb.InvoiceUpdateDesc, error) {
+
+			var (
+				updateDesc *channeldb.InvoiceUpdateDesc
+				err        error
+			)
+			updateDesc, result, err = updateInvoice(&updateCtx, inv)
+
+			// Only send an update if the invoice state was changed.
+			updateSubscribers = err == nil && updateDesc != nil &&
```
[the remark this answers is not in the document]
Comment (joostjager) [reply invoices/invoiceregistry.go]: Yes, I had the same thought.

File: invoices/invoiceregistry.go [hunk 24]
```diff
@@ -1043,3 +975,152 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
...
+	// If the invoice is already canceled, there is no further checking to
+	// do.
+	if inv.State == channeldb.ContractCanceled {
+		return nil, resultInvoiceAlreadyCanceled, nil
+	}
+
+	// If an invoice amount is specified, check that enough is paid. Also
+	// check this for duplicate payments if the invoice is already settled
+	// or accepted.
+	if inv.Terms.Value > 0 && ctx.amtPaid < inv.Terms.Value {
```
Comment (cfromknecht) [inline invoices/invoiceregistry.go]: similar to my comment on mpp, i'm not sure checking `inv.Terms.Value > 0` is necessary. i know this is preexisting but thought i'd bring it up now

File: channeldb/invoices.go [hunk 25]
```diff
@@ -1332,7 +1345,119 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
...
+	if update.NewState == ContractOpen {
+		return ErrInvoiceCannotOpen
+	}
+
+	switch invoice.State {
+
+	// Once a contract is accepted, we can only transition to settled or
+	// canceled. Forbid transitioning back into this state, otherwise this
+	// state is identical to ContractOpen, so we fallthrough to apply the
+	// same checks for accepted invoices.
```
Comment (cfromknecht) [inline channeldb/invoices.go]: i know i typed this, but "accepted" should be "open" invoices 😂

File: invoices/invoiceregistry.go (hunk 24 above)
Comment (joostjager) [reply invoices/invoiceregistry.go]: Fixed

File: channeldb/invoices.go (hunk 25 above)
Comment (joostjager) [reply channeldb/invoices.go]: fixed

File: channeldb/invoices.go [hunk 26]
```diff
@@ -1241,79 +1260,79 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
...
-
-			continue
 		}
+	}
 
-		// Add new htlc paying to the invoice.
-		if ok {
-			return nil, fmt.Errorf("htlc %v already exists", key)
+	// Process add actions from update descriptor.
+	for key, htlcUpdate := range update.AddHtlcs {
```
Comment (halseth) [inline channeldb/invoices.go]: what's the rationale behind doing the invoice state update before adding/canceling HTLCs? (which is a change of order)

File: channeldb/invoices.go (hunk 10 above)
Comment (joostjager) [reply channeldb/invoices.go]: Yes, it is a bit more strict like this. Also a matter of personal pref.

File: channeldb/invoices.go (hunk 17 above)
Comment (joostjager) [reply channeldb/invoices.go]: Another check added :)

File: invoices/invoiceregistry.go [hunk 27]
```diff
@@ -1043,3 +975,152 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
...
+		update.State = channeldb.ContractAccepted
+		return &update, resultDuplicateToAccepted, nil
+
+	case channeldb.ContractSettled:
+		update.State = channeldb.ContractSettled
+		return &update, resultDuplicateToSettled, nil
+	}
+
+	holdInvoice := inv.Terms.PaymentPreimage == channeldb.UnknownPreimage
+	if holdInvoice {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply invoices/invoiceregistry.go]: indeed. added again

File: channeldb/invoices.go [hunk 28]
```diff
@@ -1241,79 +1260,79 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
 		return &invoice, err
 	}
 
-	// Update invoice state.
-	invoice.State = update.State
+	// If there is nothing to update, return early.
+	if update == nil {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply channeldb/invoices.go]: Everywhere we return `nil` in `InvoiceRegistry.updateInvoice`. For example an htlc that has the wrong amount. We execute the update logic and conclude we don't need to update the invoice.

File: invoices/invoiceregistry.go [hunk 29]
```diff
@@ -439,111 +436,46 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
...
+
+			return updateDesc, err
+		},
+	)
+	if err != nil {
 		debugLog(err.Error())
 
 		return nil, err
 	}
+	debugLog(result.String())
```
[the remark this answers is not in the document]
Comment (joostjager) [reply invoices/invoiceregistry.go]: Not really, because an error may also occur after the callback, when writing back the invoice.

File: channeldb/invoices.go (hunk 26 above)
Comment (joostjager) [reply channeldb/invoices.go]: This isn't changing is this commit, or is it? It only pulls apart the add and cancel actions?

Comment (joostjager) [reply channeldb/invoices.go]: Or maybe you mean the next commit? But there we still update the state first, as we do in master?

File: channeldb/invoices.go [hunk 30]
```diff
@@ -1248,78 +1265,74 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
...
+			continue
 		}
 
-		err := setSettleFields(settleIndex, invoiceNum, &invoice, now)
+		err := updateHtlc(now, htlc, invoice.State)
 		if err != nil {
 			return nil, err
 		}
+
+		if invoice.State != ContractOpen &&
```
[the remark this answers is not in the document]
Comment (joostjager) [reply channeldb/invoices.go]: Comments added

File: invoices/invoiceregistry.go [hunk 31]
```diff
@@ -587,39 +589,13 @@ func (i *InvoiceRegistry) CancelInvoice(payHash lntypes.Hash) error {
...
-
-			canceledHtlcs[key] = struct{}{}
-		}
-
-		// Move invoice to the canceled state.
+		// Move invoice to the canceled state. Rely on validation in
+		// channeldb to return an error if the invoice is already
+		// settled or canceled.
 		return &channeldb.InvoiceUpdateDesc{
-			CancelHtlcs: canceledHtlcs,
```
[the remark this answers is not in the document]
Comment (joostjager) [reply invoices/invoiceregistry.go]: `channeldb.updateHtlc` does it when aligning the htlc state with the invoice state

File: invoices/update.go [hunk 32]
```diff
@@ -107,8 +107,9 @@ func updateInvoice(ctx *invoiceUpdateCtx, inv *channeldb.Invoice) (
 
 	// If an invoice amount is specified, check that enough is paid. Also
 	// check this for duplicate payments if the invoice is already settled
-	// or accepted.
-	if inv.Terms.Value > 0 && ctx.amtPaid < inv.Terms.Value {
+	// or accepted. In case this is a zero-valued invoice, it will always be
+	// enough.
+	if ctx.amtPaid < inv.Terms.Value {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply invoices/update.go]: an htlc cannot have amount 0

File: invoices/invoiceregistry.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (halseth) [reply invoices/invoiceregistry.go]: Ok, I see. We make the implicit assumption that canceling an invoice will also cancel all the HTLCs.

So `CancelHtlcs` is not really being used anymore?

File: invoices/invoiceregistry.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply invoices/invoiceregistry.go]: Yes, previously we'd log the update result even if the invoice update failed. So strictly speaking it is a change indeed.

File: channeldb/invoices.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply channeldb/invoices.go]: It is documented on the field description of `InvoiceUpdateDesc`

File: invoices/invoiceregistry.go [hunk 33]
```diff
@@ -977,154 +977,3 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
 
 	delete(i.hodlReverseSubscriptions, subscriber)
 }
-
```
[the remark this answers is not in the document]
Comment (joostjager) [reply invoices/invoiceregistry.go]: the idea is that this would make git smarter with rebases, but not sure if it actually is smarter.

File: channeldb/invoices.go (hunk 26 above)
Comment (joostjager) [reply channeldb/invoices.go]: Yes, the rationale is that we don't need to loop through the htlcs twice. So the invoice state is leading and that is set first. Then we go through the htlcs and update their states based on the invoice state and what we got from the update descriptor

File: invoices/invoiceregistry.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply invoices/invoiceregistry.go]: It is used to cancel individual htlcs in the mpp follow up.

File: channeldb/invoices.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (halseth) [reply channeldb/invoices.go]: That is for the `State` field, not the `InvoiceStateUpdateDesc` itself, which is what we check here. I think it would make sense for the callback function documentation to note this.
